Failed tests (10663): TestQuery

The following test appears to have failed:

[#10663](https://circleci.com/gh/cockroachdb/cockroach/10663):

```
I0106 10:00:24.706558 1019 storage/replica.go:1618  gossiping system config from store 1, range 1
E0106 10:00:24.707436 1019 config/config.go:336  unable to determine largest reserved object ID from system config: descriptor table not found in system config of 0 values
I0106 10:00:24.708009 1019 storage/replica.go:1567  gossiping sentinel from store 1, range 1
I0106 10:00:24.708310 1019 storage/replica.go:1573  gossiping first range from store 1, range 1
I0106 10:00:24.711049 1019 storage/replica.go:1112  reproposing 1 commands after empty entry
--- FAIL: TestQuery (0.03s)
	<autogenerated>:32: Found unexpected differences in model data and actual data:
		
		Key test.metric/@0, r:999 differs between model and actual:
		Actual value: start_timestamp_nanos:0 sample_duration_nanos:1 samples:<offset:1 count:2 sum:200 max:100 min:100 > samples:<offset:5 count:2 sum:400 max:200 min:200 > 
		Model value: start_timestamp_nanos:0 sample_duration_nanos:1 samples:<offset:1 count:1 sum:100 > samples:<offset:5 count:1 sum:200 > 
		Key test.metric/@10, r:999 differs between model and actual:
		Actual value: start_timestamp_nanos:10 sample_duration_nanos:1 samples:<offset:5 count:2 sum:600 max:300 min:300 > samples:<offset:6 count:2 sum:800 max:400 min:400 > samples:<offset:7 count:2 sum:1000 max:500 min:500 > 
		Model value: start_timestamp_nanos:10 sample_duration_nanos:1 samples:<offset:5 count:1 sum:300 > samples:<offset:6 count:1 sum:400 > samples:<offset:7 count:1 sum:500 > 
		Key test.metric/@20, r:999 differs between model and actual:
		Actual value: start_timestamp_nanos:20 sample_duration_nanos:1 samples:<offset:2 count:2 sum:1200 max:600 min:600 > 
		Model value: start_timestamp_nanos:20 sample_duration_nanos:1 samples:<offset:2 count:1 sum:600 > 
		Key test.metric/@50, r:999 differs between model and actual:
		Actual value: start_timestamp_nanos:50 sample_duration_nanos:1 samples:<offset:2 count:2 sum:1800 max:900 min:900 > 
		Model value: start_timestamp_nanos:50 sample_duration_nanos:1 samples:<offset:2 count:1 sum:900 > 
=== RUN   TestToInternal
--- FAIL: TestToInternal (0.14s)
	<autogenerated>:31: Test appears to have leaked a store:
		github.com/cockroachdb/cockroach/kv.(*TxnCoordSender).startStats(0xc820132140)
			/go/src/github.com/cockroachdb/cockroach/kv/txn_coord_sender.go:214 +0xf1a
		github.com/cockroachdb/cockroach/kv.(*TxnCoordSender).(github.com/cockroachdb/cockroach/kv.startStats)-fm()
			/go/src/github.com/cockroachdb/cockroach/kv/txn_coord_sender.go:201 +0x2e
		github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc820252070, 0xc8204d61b0)
			/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:89 +0x60
		created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
			/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:90 +0x70
		
--
RdRg@0.000000000,0  127.0.0.1:0  10:00:25.282919  48.116µs    ·meta descriptor lookup       kv/dist_sender.go:562
RdRg@0.000000000,0  127.0.0.1:0  10:00:25.282996  2.017336ms  ·sending RPC                  kv/dist_sender.go:422
RdRg@0.000000000,0  127.0.0.1:0  10:00:25.283089  0           ··sending to 127.0.0.1:51957  rpc/send.go:154

I0106 10:00:25.290749 1019 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
--- FAIL: TestHttpQuery (0.54s)
	<autogenerated>:31: Test appears to have leaked a store:
		github.com/cockroachdb/cockroach/kv.(*TxnCoordSender).startStats(0xc820132140)
			/go/src/github.com/cockroachdb/cockroach/kv/txn_coord_sender.go:214 +0xf1a
		github.com/cockroachdb/cockroach/kv.(*TxnCoordSender).(github.com/cockroachdb/cockroach/kv.startStats)-fm()
			/go/src/github.com/cockroachdb/cockroach/kv/txn_coord_sender.go:201 +0x2e
		github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc820252070, 0xc8204d61b0)
			/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:89 +0x60
		created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
			/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:90 +0x70
		
--
			/go/src/github.com/cockroachdb/cockroach/util/leaktest/leaktest.go:34 +0x2f
		github.com/cockroachdb/cockroach/ts_test.TestMain(0xc8204c3f10)
			/go/src/github.com/cockroachdb/cockroach/ts/main_test.go:34 +0x2f
		main.main()
			github.com/cockroachdb/cockroach/ts/_test/_testmain.go:70 +0x20a
FAIL
FAIL	github.com/cockroachdb/cockroach/ts	0.902s
?   	github.com/cockroachdb/cockroach/ui	[no test files]
=== RUN   TestTruncateDuration
--- PASS: TestTruncateDuration (0.00s)
=== RUN   TestErrorf
--- PASS: TestErrorf (0.00s)
=== RUN   TestErrorfSkipFrames
--- PASS: TestErrorfSkipFrames (0.00s)
=== RUN   TestFeed
--- PASS: TestFeed (0.00s)
=== RUN   TestMultipleFeed

```
Please assign, take a look and update the issue accordingly.

--- a/storage/engine/mvcc.go
+++ b/storage/engine/mvcc.go
@@ -227,8 +227,9 @@ func updateStatsForInline(ms *MVCCStats, key roachpb.Key, origMetaKeySize, origM
 // values. Unfortunately, we're unable to keep accurate stats on merge
 // as the actual details of the merge play out asynchronously during
 // compaction. Instead, we undercount by adding only the size of the
-// value.Bytes byte slice. These errors are corrected during splits
-// and merges.
+// value.Bytes byte slice (an estimated 12 bytes for timestamp,
+// included in valSize by caller). These errors are corrected during
+// splits and merges.
 func updateStatsOnMerge(ms *MVCCStats, key roachpb.Key, valSize int64) {
 	ok, sys := updateStatsForKey(ms, key)
 	if !ok {
@@ -965,7 +966,7 @@ func MVCCConditionalPut(engine Engine, ms *MVCCStats, key roachpb.Key, timestamp
 // concatenates undifferentiated byte slice values, and efficiently
 // combines time series observations if the roachpb.Value tag value
 // indicates the value byte slice is of type TIMESERIES.
-func MVCCMerge(engine Engine, ms *MVCCStats, key roachpb.Key, value roachpb.Value) error {
+func MVCCMerge(engine Engine, ms *MVCCStats, key roachpb.Key, timestamp roachpb.Timestamp, value roachpb.Value) error {
 	if len(key) == 0 {
 		return emptyKeyError()
 	}
@@ -973,6 +974,10 @@ func MVCCMerge(engine Engine, ms *MVCCStats, key roachpb.Key, value roachpb.Valu
 
 	// Encode and merge the MVCC metadata with inlined value.
 	meta := &MVCCMetadata{RawBytes: value.RawBytes}
+	// If non-zero, set the merge timestamp to provide some replay protection.
+	if !timestamp.Equal(roachpb.ZeroTimestamp) {
+		meta.MergeTimestamp = &timestamp
+	}
 	data, err := proto.Marshal(meta)
 	if err != nil {
 		return err
@@ -981,7 +986,7 @@ func MVCCMerge(engine Engine, ms *MVCCStats, key roachpb.Key, value roachpb.Valu
 		return err
 	}
 	// Every type flows through here, so we can't use the typed getters.
-	updateStatsOnMerge(ms, key, int64(len(value.RawBytes)))
+	updateStatsOnMerge(ms, key, int64(len(value.RawBytes))+mvccVersionTimestampSize)
 	return nil
 }
 

--- a/storage/engine/mvcc.pb.go
+++ b/storage/engine/mvcc.pb.go
@@ -31,7 +31,10 @@ var _ = math.Inf
 // MVCCMetadata holds MVCC metadata for a key. Used by storage/engine/mvcc.go.
 type MVCCMetadata struct {
 	Txn *cockroach_roachpb1.Transaction `protobuf:"bytes,1,opt,name=txn" json:"txn,omitempty"`
-	// The timestamp of the most recent versioned value.
+	// The timestamp of the most recent versioned value if this is a
+	// value that may have multiple versions. For values which may have
+	// only one version, the data is stored inline (via raw_bytes), and
+	// timestamp is set to zero.
 	Timestamp cockroach_roachpb1.Timestamp `protobuf:"bytes,2,opt,name=timestamp" json:"timestamp"`
 	// Is the most recent value a deletion tombstone?
 	Deleted bool `protobuf:"varint,3,opt,name=deleted" json:"deleted"`
@@ -39,12 +42,16 @@ type MVCCMetadata struct {
 	KeyBytes int64 `protobuf:"varint,4,opt,name=key_bytes" json:"key_bytes"`
 	// The size in bytes of the most recent versioned value.
 	ValBytes int64 `protobuf:"varint,5,opt,name=val_bytes" json:"val_bytes"`
-	// Inline value, used for values with zero timestamp. This provides
-	// an efficient short circuit of the normal MVCC metadata sentinel
-	// and subsequent version rows. If timestamp == (0, 0), then there
-	// is only a single MVCC metadata row with value inlined, and with
-	// empty timestamp, key_bytes, and val_bytes.
+	// Inline value, used for non-versioned values with zero
+	// timestamp. This provides an efficient short circuit of the normal
+	// MVCC metadata sentinel and subsequent version rows. If timestamp
+	// == (0, 0), then there is only a single MVCC metadata row with
+	// value inlined, and with empty timestamp, key_bytes, and
+	// val_bytes.
 	RawBytes []byte `protobuf:"bytes,6,opt,name=raw_bytes" json:"raw_bytes,omitempty"`
+	// This provides a measure of protection against replays caused by
+	// Raft duplicating merge commands.
+	MergeTimestamp *cockroach_roachpb1.Timestamp `protobuf:"bytes,7,opt,name=merge_timestamp" json:"merge_timestamp,omitempty"`
 }
 
 func (m *MVCCMetadata) Reset()         { *m = MVCCMetadata{} }
@@ -137,6 +144,16 @@ func (m *MVCCMetadata) MarshalTo(data []byte) (int, error) {
 		i++
 		i = encodeVarintMvcc(data, i, uint64(len(m.RawBytes)))
 		i += copy(data[i:], m.RawBytes)
+	}
+	if m.MergeTimestamp != nil {
+		data[i] = 0x3a
+		i++
+		i = encodeVarintMvcc(data, i, uint64(m.MergeTimestamp.Size()))
+		n3, err := m.MergeTimestamp.MarshalTo(data[i:])
+		if err != nil {
+			return 0, err
+		}
+		i += n3
 	}
 	return i, nil
 }
@@ -241,6 +258,10 @@ func (m *MVCCMetadata) Size() (n int) {
 	n += 1 + sovMvcc(uint64(m.ValBytes))
 	if m.RawBytes != nil {
 		l = len(m.RawBytes)
+		n += 1 + l + sovMvcc(uint64(l))
+	}
+	if m.MergeTimestamp != nil {
+		l = m.MergeTimestamp.Size()
 		n += 1 + l + sovMvcc(uint64(l))
 	}
 	return n
@@ -455,6 +476,39 @@ func (m *MVCCMetadata) Unmarshal(data []byte) error {
 				return io.ErrUnexpectedEOF
 			}
 			m.RawBytes = append([]byte{}, data[iNdEx:postIndex]...)
+			iNdEx = postIndex
+		case 7:
+			if wireType != 2 {
+				return fmt.Errorf("proto: wrong wireType = %d for field MergeTimestamp", wireType)
+			}
+			var msglen int
+			for shift := uint(0); ; shift += 7 {
+				if shift >= 64 {
+					return ErrIntOverflowMvcc
+				}
+				if iNdEx >= l {
+					return io.ErrUnexpectedEOF
+				}
+				b := data[iNdEx]
+				iNdEx++
+				msglen |= (int(b) & 0x7F) << shift
+				if b < 0x80 {
+					break
+				}
+			}
+			if msglen < 0 {
+				return ErrInvalidLengthMvcc
+			}
+			postIndex := iNdEx + msglen
+			if postIndex > l {
+				return io.ErrUnexpectedEOF
+			}
+			if m.MergeTimestamp == nil {
+				m.MergeTimestamp = &cockroach_roachpb1.Timestamp{}
+			}
+			if err := m.MergeTimestamp.Unmarshal(data[iNdEx:postIndex]); err != nil {
+				return err
+			}
 			iNdEx = postIndex
 		default:
 			iNdEx = preIndex

--- a/storage/engine/mvcc.proto
+++ b/storage/engine/mvcc.proto
@@ -24,7 +24,10 @@ import weak "gogoproto/gogo.proto";
 // MVCCMetadata holds MVCC metadata for a key. Used by storage/engine/mvcc.go.
 message MVCCMetadata {
   optional roachpb.Transaction txn = 1;
-  // The timestamp of the most recent versioned value.
+  // The timestamp of the most recent versioned value if this is a
+  // value that may have multiple versions. For values which may have
+  // only one version, the data is stored inline (via raw_bytes), and
+  // timestamp is set to zero.
   optional roachpb.Timestamp timestamp = 2 [(gogoproto.nullable) = false];
   // Is the most recent value a deletion tombstone?
   optional bool deleted = 3 [(gogoproto.nullable) = false];
@@ -32,12 +35,16 @@ message MVCCMetadata {
   optional int64 key_bytes = 4 [(gogoproto.nullable) = false];
   // The size in bytes of the most recent versioned value.
   optional int64 val_bytes = 5 [(gogoproto.nullable) = false];
-  // Inline value, used for values with zero timestamp. This provides
-  // an efficient short circuit of the normal MVCC metadata sentinel
-  // and subsequent version rows. If timestamp == (0, 0), then there
-  // is only a single MVCC metadata row with value inlined, and with
-  // empty timestamp, key_bytes, and val_bytes.
+  // Inline value, used for non-versioned values with zero
+  // timestamp. This provides an efficient short circuit of the normal
+  // MVCC metadata sentinel and subsequent version rows. If timestamp
+  // == (0, 0), then there is only a single MVCC metadata row with
+  // value inlined, and with empty timestamp, key_bytes, and
+  // val_bytes.
   optional bytes raw_bytes = 6;
+  // This provides a measure of protection against replays caused by
+  // Raft duplicating merge commands.
+  optional roachpb.Timestamp merge_timestamp = 7;
 }
 
 // MVCCStats tracks byte and instance counts for:

--- a/storage/engine/rocksdb/cockroach/storage/engine/mvcc.pb.cc
+++ b/storage/engine/rocksdb/cockroach/storage/engine/mvcc.pb.cc
@@ -40,13 +40,14 @@ void protobuf_AssignDesc_cockroach_2fstorage_2fengine_2fmvcc_2eproto() {
       "cockroach/storage/engine/mvcc.proto");
   GOOGLE_CHECK(file != NULL);
   MVCCMetadata_descriptor_ = file->message_type(0);
-  static const int MVCCMetadata_offsets_[6] = {
+  static const int MVCCMetadata_offsets_[7] = {
     GOOGLE_PROTOBUF_GENERATED_MESSAGE_FIELD_OFFSET(MVCCMetadata, txn_),
     GOOGLE_PROTOBUF_GENERATED_MESSAGE_FIELD_OFFSET(MVCCMetadata, timestamp_),
     GOOGLE_PROTOBUF_GENERATED_MESSAGE_FIELD_OFFSET(MVCCMetadata, deleted_),
     GOOGLE_PROTOBUF_GENERATED_MESSAGE_FIELD_OFFSET(MVCCMetadata, key_bytes_),
     GOOGLE_PROTOBUF_GENERATED_MESSAGE_FIELD_OFFSET(MVCCMetadata, val_bytes_),
     GOOGLE_PROTOBUF_GENERATED_MESSAGE_FIELD_OFFSET(MVCCMetadata, raw_bytes_),
+    GOOGLE_PROTOBUF_GENERATED_MESSAGE_FIELD_OFFSET(MVCCMetadata, merge_timestamp_),
   };
   MVCCMetadata_reflection_ =
     ::google::protobuf::internal::GeneratedMessageReflection::NewGeneratedMessageReflection(
@@ -124,22 +125,24 @@ void protobuf_AddDesc_cockroach_2fstorage_2fengine_2fmvcc_2eproto() {
   ::google::protobuf::DescriptorPool::InternalAddGeneratedFile(
     "\n#cockroach/storage/engine/mvcc.proto\022\030c"
     "ockroach.storage.engine\032\034cockroach/roach"
-    "pb/data.proto\032\024gogoproto/gogo.proto\"\316\001\n\014"
+    "pb/data.proto\032\024gogoproto/gogo.proto\"\205\002\n\014"
     "MVCCMetadata\022+\n\003txn\030\001 \001(\0132\036.cockroach.ro"
     "achpb.Transaction\0225\n\ttimestamp\030\002 \001(\0132\034.c"
     "ockroach.roachpb.TimestampB\004\310\336\037\000\022\025\n\007dele"
     "ted\030\003 \001(\010B\004\310\336\037\000\022\027\n\tkey_bytes\030\004 \001(\003B\004\310\336\037\000"
     "\022\027\n\tval_bytes\030\005 \001(\003B\004\310\336\037\000\022\021\n\traw_bytes\030\006"
-    " \001(\014\"\362\002\n\tMVCCStats\022\030\n\nlive_bytes\030\001 \001(\003B\004"
-    "\310\336\037\000\022\027\n\tkey_bytes\030\002 \001(\003B\004\310\336\037\000\022\027\n\tval_byt"
-    "es\030\003 \001(\003B\004\310\336\037\000\022\032\n\014intent_bytes\030\004 \001(\003B\004\310\336"
-    "\037\000\022\030\n\nlive_count\030\005 \001(\003B\004\310\336\037\000\022\027\n\tkey_coun"
-    "t\030\006 \001(\003B\004\310\336\037\000\022\027\n\tval_count\030\007 \001(\003B\004\310\336\037\000\022\032"
-    "\n\014intent_count\030\010 \001(\003B\004\310\336\037\000\022\030\n\nintent_age"
-    "\030\t \001(\003B\004\310\336\037\000\022(\n\014gc_bytes_age\030\n \001(\003B\022\310\336\037\000"
-    "\342\336\037\nGCBytesAge\022\027\n\tsys_bytes\030\014 \001(\003B\004\310\336\037\000\022"
-    "\027\n\tsys_count\030\r \001(\003B\004\310\336\037\000\022\037\n\021last_update_"
-    "nanos\030\036 \001(\003B\004\310\336\037\000B\010Z\006engineX\001", 709);
+    " \001(\014\0225\n\017merge_timestamp\030\007 \001(\0132\034.cockroac"
+    "h.roachpb.Timestamp\"\362\002\n\tMVCCStats\022\030\n\nliv"
+    "e_bytes\030\001 \001(\003B\004\310\336\037\000\022\027\n\tkey_bytes\030\002 \001(\003B\004"
+    "\310\336\037\000\022\027\n\tval_bytes\030\003 \001(\003B\004\310\336\037\000\022\032\n\014intent_"
+    "bytes\030\004 \001(\003B\004\310\336\037\000\022\030\n\nlive_count\030\005 \001(\003B\004\310"
+    "\336\037\000\022\027\n\tkey_count\030\006 \001(\003B\004\310\336\037\000\022\027\n\tval_coun"
+    "t\030\007 \001(\003B\004\310\336\037\000\022\032\n\014intent_count\030\010 \001(\003B\004\310\336\037"
+    "\000\022\030\n\nintent_age\030\t \001(\003B\004\310\336\037\000\022(\n\014gc_bytes_"
+    "age\030\n \001(\003B\022\310\336\037\000\342\336\037\nGCBytesAge\022\027\n\tsys_byt"
+    "es\030\014 \001(\003B\004\310\336\037\000\022\027\n\tsys_count\030\r \001(\003B\004\310\336\037\000\022"
+    "\037\n\021last_update_nanos\030\036 \001(\003B\004\310\336\037\000B\010Z\006engi"
+    "neX\001", 764);
   ::google::protobuf::MessageFactory::InternalRegisterGeneratedFile(
     "cockroach/storage/engine/mvcc.proto", &protobuf_RegisterTypes);
   MVCCMetadata::default_instance_ = new MVCCMetadata();
@@ -175,6 +178,7 @@ const int MVCCMetadata::kDeletedFieldNumber;
 const int MVCCMetadata::kKeyBytesFieldNumber;
 const int MVCCMetadata::kValBytesFieldNumber;
 const int MVCCMetadata::kRawBytesFieldNumber;
+const int MVCCMetadata::kMergeTimestampFieldNumber;
 #endif  // !defined(_MSC_VER) || _MSC_VER >= 1900
 
 MVCCMetadata::MVCCMetadata()
@@ -186,6 +190,7 @@ MVCCMetadata::MVCCMetadata()
 void MVCCMetadata::InitAsDefaultInstance() {
   txn_ = const_cast< ::cockroach::roachpb::Transaction*>(&::cockroach::roachpb::Transaction::default_instance());
   timestamp_ = const_cast< ::cockroach::roachpb::Timestamp*>(&::cockroach::roachpb::Timestamp::default_instance());
+  merge_timestamp_ = const_cast< ::cockroach::roachpb::Timestamp*>(&::cockroach::roachpb::Timestamp::default_instance());
 }
 
 MVCCMetadata::MVCCMetadata(const MVCCMetadata& from)
@@ -205,6 +210,7 @@ void MVCCMetadata::SharedCtor() {
   key_bytes_ = GOOGLE_LONGLONG(0);
   val_bytes_ = GOOGLE_LONGLONG(0);
   raw_bytes_.UnsafeSetDefault(&::google::protobuf::internal::GetEmptyStringAlreadyInited());
+  merge_timestamp_ = NULL;
   ::memset(_has_bits_, 0, sizeof(_has_bits_));
 }
 
@@ -218,6 +224,7 @@ void MVCCMetadata::SharedDtor() {
   if (this != default_instance_) {
     delete txn_;
     delete timestamp_;
+    delete merge_timestamp_;
   }
 }
 
@@ -255,7 +262,7 @@ void MVCCMetadata::Clear() {
            ZR_HELPER_(last) - ZR_HELPER_(first) + sizeof(last));\
 } while (0)
 
-  if (_has_bits_[0 / 32] & 63u) {
+  if (_has_bits_[0 / 32] & 127u) {
     ZR_(key_bytes_, val_bytes_);
     if (has_txn()) {
       if (txn_ != NULL) txn_->::cockroach::roachpb::Transaction::Clear();
@@ -266,6 +273,9 @@ void MVCCMetadata::Clear() {
     deleted_ = false;
     if (has_raw_bytes()) {
       raw_bytes_.ClearToEmptyNoArena(&::google::protobuf::internal::GetEmptyStringAlreadyInited());
+    }
+    if (has_merge_timestamp()) {
+      if (merge_timestamp_ != NULL) merge_timestamp_->::cockroach::roachpb::Timestamp::Clear();
     }
   }
 
@@ -367,6 +377,19 @@ bool MVCCMetadata::MergePartialFromCodedStream(
         } else {
           goto handle_unusual;
         }
+        if (input->ExpectTag(58)) goto parse_merge_timestamp;
+        break;
+      }
+
+      // optional .cockroach.roachpb.Timestamp merge_timestamp = 7;
+      case 7: {
+        if (tag == 58) {
+         parse_merge_timestamp:
+          DO_(::google::protobuf::internal::WireFormatLite::ReadMessageNoVirtual(
+               input, mutable_merge_timestamp()));
+        } else {
+          goto handle_unusual;
+        }
         if (input->ExpectAtEnd()) goto success;
         break;
       }
@@ -429,6 +452,12 @@ void MVCCMetadata::SerializeWithCachedSizes(
       6, this->raw_bytes(), output);
   }
 
+  // optional .cockroach.roachpb.Timestamp merge_timestamp = 7;
+  if (has_merge_timestamp()) {
+    ::google::protobuf::internal::WireFormatLite::WriteMessageMaybeToArray(
+      7, *this->merge_timestamp_, output);
+  }
+
   if (_internal_metadata_.have_unknown_fields()) {
     ::google::protobuf::internal::WireFormat::SerializeUnknownFields(
         unknown_fields(), output);
@@ -475,6 +504,13 @@ void MVCCMetadata::SerializeWithCachedSizes(
         6, this->raw_bytes(), target);
   }
 
+  // optional .cockroach.roachpb.Timestamp merge_timestamp = 7;
+  if (has_merge_timestamp()) {
+    target = ::google::protobuf::internal::WireFormatLite::
+      WriteMessageNoVirtualToArray(
+        7, *this->merge_timestamp_, target);
+  }
+
   if (_internal_metadata_.have_unknown_fields()) {
     target = ::google::protobuf::internal::WireFormat::SerializeUnknownFieldsToArray(
         unknown_fields(), target);
@@ -486,7 +522,7 @@ void MVCCMetadata::SerializeWithCachedSizes(
 int MVCCMetadata::ByteSize() const {
   int total_size = 0;
 
-  if (_has_bits_[0 / 32] & 63u) {
+  if (_has_bits_[0 / 32] & 127u) {
     // optional .cockroach.roachpb.Transaction txn = 1;
     if (has_txn()) {
       total_size += 1 +
@@ -525,6 +561,13 @@ int MVCCMetadata::ByteSize() const {
       total_size += 1 +
         ::google::protobuf::internal::WireFormatLite::BytesSize(
           this->raw_bytes());
+    }
+
+    // optional .cockroach.roachpb.Timestamp merge_timestamp = 7;
+    if (has_merge_timestamp()) {
+      total_size += 1 +
+        ::google::protobuf::internal::WireFormatLite::MessageSizeNoVirtual(
+          *this->merge_timestamp_);
     }
 
   }
@@ -573,6 +616,9 @@ void MVCCMetadata::MergeFrom(const MVCCMetadata& from) {
       set_has_raw_bytes();
       raw_bytes_.AssignWithDefault(&::google::protobuf::internal::GetEmptyStringAlreadyInited(), from.raw_bytes_);
     }
+    if (from.has_merge_timestamp()) {
+      mutable_merge_timestamp()->::cockroach::roachpb::Timestamp::MergeFrom(from.merge_timestamp());
+    }
   }
   if (from._internal_metadata_.have_unknown_fields()) {
     mutable_unknown_fields()->MergeFrom(from.unknown_fields());
@@ -607,6 +653,7 @@ void MVCCMetadata::InternalSwap(MVCCMetadata* other) {
   std::swap(key_bytes_, other->key_bytes_);
   std::swap(val_bytes_, other->val_bytes_);
   raw_bytes_.Swap(&other->raw_bytes_);
+  std::swap(merge_timestamp_, other->merge_timestamp_);
   std::swap(_has_bits_[0], other->_has_bits_[0]);
   _internal_metadata_.Swap(&other->_internal_metadata_);
   std::swap(_cached_size_, other->_cached_size_);
@@ -832,6 +879,49 @@ void MVCCMetadata::clear_raw_bytes() {
   }
   raw_bytes_.SetAllocatedNoArena(&::google::protobuf::internal::GetEmptyStringAlreadyInited(), raw_bytes);
   // @@protoc_insertion_point(field_set_allocated:cockroach.storage.engine.MVCCMetadata.raw_bytes)
+}
+
+// optional .cockroach.roachpb.Timestamp merge_timestamp = 7;
+bool MVCCMetadata::has_merge_timestamp() const {
+  return (_has_bits_[0] & 0x00000040u) != 0;
+}
+void MVCCMetadata::set_has_merge_timestamp() {
+  _has_bits_[0] |= 0x00000040u;
+}
+void MVCCMetadata::clear_has_merge_timestamp() {
+  _has_bits_[0] &= ~0x00000040u;
+}
+void MVCCMetadata::clear_merge_timestamp() {
+  if (merge_timestamp_ != NULL) merge_timestamp_->::cockroach::roachpb::Timestamp::Clear();
+  clear_has_merge_timestamp();
+}
+const ::cockroach::roachpb::Timestamp& MVCCMetadata::merge_timestamp() const {
+  // @@protoc_insertion_point(field_get:cockroach.storage.engine.MVCCMetadata.merge_timestamp)
+  return merge_timestamp_ != NULL ? *merge_timestamp_ : *default_instance_->merge_timestamp_;
+}
+::cockroach::roachpb::Timestamp* MVCCMetadata::mutable_merge_timestamp() {
+  set_has_merge_timestamp();
+  if (merge_timestamp_ == NULL) {
+    merge_timestamp_ = new ::cockroach::roachpb::Timestamp;
+  }
+  // @@protoc_insertion_point(field_mutable:cockroach.storage.engine.MVCCMetadata.merge_timestamp)
+  return merge_timestamp_;
+}
+::cockroach::roachpb::Timestamp* MVCCMetadata::release_merge_timestamp() {
+  clear_has_merge_timestamp();
+  ::cockroach::roachpb::Timestamp* temp = merge_timestamp_;
+  merge_timestamp_ = NULL;
+  return temp;
+}
+void MVCCMetadata::set_allocated_merge_timestamp(::cockroach::roachpb::Timestamp* merge_timestamp) {
+  delete merge_timestamp_;
+  merge_timestamp_ = merge_timestamp;
+  if (merge_timestamp) {
+    set_has_merge_timestamp();
+  } else {
+    clear_has_merge_timestamp();
+  }
+  // @@protoc_insertion_point(field_set_allocated:cockroach.storage.engine.MVCCMetadata.merge_timestamp)
 }
 
 #endif  // PROTOBUF_INLINE_NOT_IN_HEADERS

--- a/storage/engine/rocksdb/cockroach/storage/engine/mvcc.pb.h
+++ b/storage/engine/rocksdb/cockroach/storage/engine/mvcc.pb.h
@@ -160,6 +160,15 @@ class MVCCMetadata : public ::google::protobuf::Message {
   ::std::string* release_raw_bytes();
   void set_allocated_raw_bytes(::std::string* raw_bytes);
 
+  // optional .cockroach.roachpb.Timestamp merge_timestamp = 7;
+  bool has_merge_timestamp() const;
+  void clear_merge_timestamp();
+  static const int kMergeTimestampFieldNumber = 7;
+  const ::cockroach::roachpb::Timestamp& merge_timestamp() const;
+  ::cockroach::roachpb::Timestamp* mutable_merge_timestamp();
+  ::cockroach::roachpb::Timestamp* release_merge_timestamp();
+  void set_allocated_merge_timestamp(::cockroach::roachpb::Timestamp* merge_timestamp);
+
   // @@protoc_insertion_point(class_scope:cockroach.storage.engine.MVCCMetadata)
  private:
   inline void set_has_txn();
@@ -174,6 +183,8 @@ class MVCCMetadata : public ::google::protobuf::Message {
   inline void clear_has_val_bytes();
   inline void set_has_raw_bytes();
   inline void clear_has_raw_bytes();
+  inline void set_has_merge_timestamp();
+  inline void clear_has_merge_timestamp();
 
   ::google::protobuf::internal::InternalMetadataWithArena _internal_metadata_;
   ::google::protobuf::uint32 _has_bits_[1];
@@ -183,6 +194,7 @@ class MVCCMetadata : public ::google::protobuf::Message {
   ::google::protobuf::int64 key_bytes_;
   ::google::protobuf::int64 val_bytes_;
   ::google::protobuf::internal::ArenaStringPtr raw_bytes_;
+  ::cockroach::roachpb::Timestamp* merge_timestamp_;
   bool deleted_;
   friend void  protobuf_AddDesc_cockroach_2fstorage_2fengine_2fmvcc_2eproto();
   friend void protobuf_AssignDesc_cockroach_2fstorage_2fengine_2fmvcc_2eproto();
@@ -617,6 +629,49 @@ inline void MVCCMetadata::set_allocated_raw_bytes(::std::string* raw_bytes) {
   }
   raw_bytes_.SetAllocatedNoArena(&::google::protobuf::internal::GetEmptyStringAlreadyInited(), raw_bytes);
   // @@protoc_insertion_point(field_set_allocated:cockroach.storage.engine.MVCCMetadata.raw_bytes)
+}
+
+// optional .cockroach.roachpb.Timestamp merge_timestamp = 7;
+inline bool MVCCMetadata::has_merge_timestamp() const {
+  return (_has_bits_[0] & 0x00000040u) != 0;
+}
+inline void MVCCMetadata::set_has_merge_timestamp() {
+  _has_bits_[0] |= 0x00000040u;
+}
+inline void MVCCMetadata::clear_has_merge_timestamp() {
+  _has_bits_[0] &= ~0x00000040u;
+}
+inline void MVCCMetadata::clear_merge_timestamp() {
+  if (merge_timestamp_ != NULL) merge_timestamp_->::cockroach::roachpb::Timestamp::Clear();
+  clear_has_merge_timestamp();
+}
+inline const ::cockroach::roachpb::Timestamp& MVCCMetadata::merge_timestamp() const {
+  // @@protoc_insertion_point(field_get:cockroach.storage.engine.MVCCMetadata.merge_timestamp)
+  return merge_timestamp_ != NULL ? *merge_timestamp_ : *default_instance_->merge_timestamp_;
+}
+inline ::cockroach::roachpb::Timestamp* MVCCMetadata::mutable_merge_timestamp() {
+  set_has_merge_timestamp();
+  if (merge_timestamp_ == NULL) {
+    merge_timestamp_ = new ::cockroach::roachpb::Timestamp;
+  }
+  // @@protoc_insertion_point(field_mutable:cockroach.storage.engine.MVCCMetadata.merge_timestamp)
+  return merge_timestamp_;
+}
+inline ::cockroach::roachpb::Timestamp* MVCCMetadata::release_merge_timestamp() {
+  clear_has_merge_timestamp();
+  ::cockroach::roachpb::Timestamp* temp = merge_timestamp_;
+  merge_timestamp_ = NULL;
+  return temp;
+}
+inline void MVCCMetadata::set_allocated_merge_timestamp(::cockroach::roachpb::Timestamp* merge_timestamp) {
+  delete merge_timestamp_;
+  merge_timestamp_ = merge_timestamp;
+  if (merge_timestamp) {
+    set_has_merge_timestamp();
+  } else {
+    clear_has_merge_timestamp();
+  }
+  // @@protoc_insertion_point(field_set_allocated:cockroach.storage.engine.MVCCMetadata.merge_timestamp)
 }
 
 // -------------------------------------------------------------------

--- a/storage/engine/rocksdb/db.cc
+++ b/storage/engine/rocksdb/db.cc
@@ -616,6 +616,25 @@ bool MergeValues(cockroach::storage::engine::MVCCMetadata *left,
       rocksdb::Warn(logger, "inconsistent value types for merge (left = bytes, right = ?)");
       return false;
     }
+    // Check for replay; if left merge_timestamp is equal to right,
+    // then ignore the merge request. Replays can happen on merges as
+    // there is no higher-level replay protection and Raft may
+    // duplicate commands. The replay protection here is imperfect.
+    // Raft may duplicate and replay commands out of order, in which
+    // case the same command may be re-merged.
+    //
+    // NOTE: there are test cases in storage/engine/merge_test and in
+    //   ts/db_test which will send MVCCMetadata with empty
+    //   merge_timestamps and should always be merged; thus the check
+    //   for non-empty merge_timestamps.
+    if (left->has_merge_timestamp() && right.has_merge_timestamp()) {
+      if (left->merge_timestamp().wall_time() == right.merge_timestamp().wall_time() &&
+          left->merge_timestamp().logical() == right.merge_timestamp().logical()) {
+        return true;
+      }
+      left->mutable_merge_timestamp()->CopyFrom(right.merge_timestamp());
+    }
+
     if (IsTimeSeriesData(left->raw_bytes()) || IsTimeSeriesData(right.raw_bytes())) {
       // The right operand must also be a time series.
       if (!IsTimeSeriesData(left->raw_bytes()) || !IsTimeSeriesData(right.raw_bytes())) {
@@ -631,6 +650,9 @@ bool MergeValues(cockroach::storage::engine::MVCCMetadata *left,
     return true;
   } else {
     left->mutable_raw_bytes()->assign(right.raw_bytes());
+    if (right.has_merge_timestamp()) {
+      left->mutable_merge_timestamp()->CopyFrom(right.merge_timestamp());
+    }
     if (full_merge && IsTimeSeriesData(left->raw_bytes())) {
       ConsolidateTimeSeriesValue(left->mutable_raw_bytes(), logger);
     }
@@ -679,7 +701,6 @@ class DBMergeOperator : public rocksdb::MergeOperator {
     // corruption error will be returned, but likely only after the next
     // read of the key). In effect, there is no propagation of error
     // information to the client.
-
     cockroach::storage::engine::MVCCMetadata meta;
     if (existing_value != NULL) {
       if (!meta.ParseFromArray(existing_value->data(), existing_value->size())) {

--- a/storage/engine/rocksdb_test.go
+++ b/storage/engine/rocksdb_test.go
@@ -455,11 +455,13 @@ func runMVCCMerge(value *roachpb.Value, numKeys int, b *testing.B) {
 
 	b.ResetTimer()
 
+	ts := roachpb.Timestamp{}
 	// Use parallelism if specified when test is run.
 	b.RunParallel(func(pb *testing.PB) {
 		for pb.Next() {
 			ms := MVCCStats{}
-			if err := MVCCMerge(rocksdb, &ms, keys[rand.Intn(numKeys)], *value); err != nil {
+			ts.Logical++
+			if err := MVCCMerge(rocksdb, &ms, keys[rand.Intn(numKeys)], ts, *value); err != nil {
 				b.Fatal(err)
 			}
 		}

--- a/storage/replica_command.go
+++ b/storage/replica_command.go
@@ -1065,14 +1065,15 @@ func (r *Replica) ResolveIntentRange(batch engine.Engine, ms *engine.MVCCStats,
 }
 
 // Merge is used to merge a value into an existing key. Merge is an
-// efficient accumulation operation which is exposed by RocksDB, used by
-// Cockroach for the efficient accumulation of certain values. Due to the
-// difficulty of making these operations transactional, merges are not currently
-// exposed directly to clients. Merged values are explicitly not MVCC data.
+// efficient accumulation operation which is exposed by RocksDB, used
+// by CockroachDB for the efficient accumulation of certain
+// values. Due to the difficulty of making these operations
+// transactional, merges are not currently exposed directly to
+// clients. Merged values are explicitly not MVCC data.
 func (r *Replica) Merge(batch engine.Engine, ms *engine.MVCCStats, h roachpb.Header, args roachpb.MergeRequest) (roachpb.MergeResponse, error) {
 	var reply roachpb.MergeResponse
 
-	return reply, engine.MVCCMerge(batch, ms, args.Key, args.Value)
+	return reply, engine.MVCCMerge(batch, ms, args.Key, h.Timestamp, args.Value)
 }
 
 // TruncateLog discards a prefix of the raft log. Truncating part of a log that

--- a/storage/replica_test.go
+++ b/storage/replica_test.go
@@ -2710,7 +2710,6 @@ func TestRangeStatsComputation(t *testing.T) {
 // that values passed to Merge are being merged.
 func TestMerge(t *testing.T) {
 	defer leaktest.AfterTest(t)
-	t.Skip("#3555")
 	tc := testContext{}
 	tc.Start(t)
 	defer tc.Stop()

--- a/ts/db_test.go
+++ b/ts/db_test.go
@@ -250,7 +250,6 @@ func datapoint(timestamp int64, val float64) *TimeSeriesDatapoint {
 // it is storing time series correctly.
 func TestStoreTimeSeries(t *testing.T) {
 	defer leaktest.AfterTest(t)
-	t.Skip("#3646")
 	tm := newTestModel(t)
 	tm.Start()
 	defer tm.Stop()

--- a/ts/query_test.go
+++ b/ts/query_test.go
@@ -311,7 +311,6 @@ func (tm *testModel) assertQuery(name string, sources []string, agg *TimeSeriesQ
 // model.
 func TestQuery(t *testing.T) {
 	defer leaktest.AfterTest(t)
-	t.Skip("#3637")
 	tm := newTestModel(t)
 	tm.Start()
 	defer tm.Stop()

